### PR TITLE
RabbitMQ

### PR DESCRIPTION
**Pull request by valentinyanakiev**
- RabbitMQ support for PubSub. Added the following:
   - RabbitMQ as part of docker-composed services. The configuration is based on bitnami's image.
   - Traefik routing for RabbitMQ's AMQP endpoit (TCP / port 5672) and management endpoint (HTTP / port 15672). RabbitMQ management interface can be accessed on http://localhost:15672
   - Replaced PubSub references (Apollo's in-memory, non-production ready PubSub implementation of PubSubEngine) with PubSubEngine. 
   - Replaced our PubSub service to use [AMQP](https://www.rabbitmq.com/tutorials/amqp-concepts.html) PubSub with RabbitMQ as AMQP provider
   - Added configuration via environment variables
   - Added extensive inline documentation in alkemio.yml for RabbitMQ's config

### CODE REVIEW

File: alkemio.yml [hunk 1]
```diff
@@ -312,6 +312,70 @@ storage:
...
+      # RabbitMQ exchange. The RabbitMQ messaging model uses producer-queue-consumer.
+      # The core idea in the messaging model in RabbitMQ is that the producer never sends any messages directly to a queue.
+      # Actually, quite often the producer doesn't even know if a message will be delivered to any queue at all.
+      # Instead, the producer can only send messages to an exchange. An exchange is a very simple thing.
+      # On one side it receives messages from producers and the other side it pushes them to queues.
+      # The exchange must know exactly what to do with a message it receives. Should it be appended to a particular queue?
+      # Should it be appended to many queues? Or should it get discarded. The rules for that are defined by the exchange type.
+      exchange:
+        # RabbitMQ exchange name
+        name: ${RABBITMQ_EXCHANGE_NAME}:alkemio-graphql-subscriptions
```
Comment (techsmyth) [inline alkemio.yml:342]: We will be using rabbitmq for a lot more I think than just graphql subscriptions - right?

File: quickstart-services.yml [hunk 2]
```diff
@@ -2,6 +2,8 @@ version: '3.7'
 
 volumes:
   alkemio_dev_mysql_db:
+  rabbitmq_data:
+    driver: local
```
Comment (techsmyth) [inline quickstart-services.yml:6]: Keen to understand options that are available for in a production setting

File: src/services/platform/matrix/agent/matrix.agent.service.ts [hunk 3]
```diff
@@ -16,7 +16,7 @@ import { MatrixRoom } from '../adapter-room/matrix.room';
 import { MatrixEntityNotFoundException } from '@common/exceptions';
 import { WINSTON_MODULE_NEST_PROVIDER } from 'nest-winston';
 import { PUB_SUB } from '@services/platform/subscription/subscription.module';
-import { PubSub } from 'graphql-subscriptions';
+import { PubSubEngine } from 'graphql-subscriptions';
```
Comment (techsmyth) [inline src/services/platform/matrix/agent/matrix.agent.service.ts:19]: What is the difference between PubSub + PubSubEngine?

File: quickstart-services.yml [hunk 4]
```diff
@@ -145,3 +149,19 @@ services:
...
+  rabbitmq:
+    container_name: alkemio_dev_rabbitmq
+    image: docker.io/bitnami/rabbitmq:3.9
+    restart: always
+    networks:
+      - alkemio_dev_net
+    ports:
+      - 4369
+      - 5672
+      - 25672
```
Comment (techsmyth) [inline quickstart-services.yml:162]: why also exposing 25672?

**Review by techsmyth — COMMENTED** [review]
Great to have this in, look like it is really nicely setup. Admin interface was working nicely for me. 

Brilliant that you sorted the breakpoints not binding issue :)

A couple of minor points - looking forward to seeing this all integrated with the subscriptions work in the community updates branch when we get them both onto develop.

File: quickstart-services.yml (hunk 2 above)
Comment (valentinyanakiev) [reply quickstart-services.yml:6]: That's not very relevant for production setting - in production there are rabbitmq kubectl plugin + rabbitmq cluster operator kubectl plugin that take care for the setup

File: alkemio.yml (hunk 1 above)
Comment (valentinyanakiev) [reply alkemio.yml:342]: This is a default, can be changed with env variable. We can change later.

File: quickstart-services.yml (hunk 4 above)
Comment (valentinyanakiev) [reply quickstart-services.yml:162]: It's not exposed outside of the network. It was the original set of ports required for the bitnami rabbitmq container.

**Review by techsmyth — APPROVED** [review]
Nice to have this in!

**Pull request merged**